Implemented Gradient for RGBToHSV

kmh4321 opened the pull request:
Hello, 

This PR implements the gradient for RGB to HSV operation which was marked as a [TODO](https://github.com/tensorflow/tensorflow/blob/08082cdab025ff095001e42839d3efbf9cd45a8e/tensorflow/python/ops/image_ops_impl.py#L46). The function is piecewise continuous and has a lot of sub-cases (with MAX() and MIN()) so the derivative calculation tends to get a bit complicated. I have left as much comments as possible in the code to explain the derivative calculations but I'd be glad to add something else if needed (i have a lot of whiteboard/pen-and-paper notes on this derivation!). 

I tried to add a test case to `tensorflow/python/ops/image_grad_test.py`, but I wasn't sure if that file is actually a real test file as I didn't find any bazel linakges. Most of the other functions in that file also seem to be using deprecated gradient functions which aren't part of TF2.0 so I wanted to first check here before I add any test(s) to the wrong file.

kmh4321 committed:
Removed not diff flag for RGB to HSV.

kmh4321 committed:
Set flags for max and min layers.

kmh4321 committed:
Derivative of rgb wrt value.

kmh4321 committed:
derivative wrt saturation.

kmh4321 committed:
dh_dr

kmh4321 committed:
dh_dg

kmh4321 committed:
dh_db

kmh4321 committed:
adding all gradients.

kmh4321 committed:
Added custom reciprocal function.

kmh4321 committed:
Fixed typo.

kmh4321 committed:
pylint fixes.

kmh4321 committed:
Updated function description.

kmh4321 committed:
Linting issue.

kmh4321 committed:
fixed import and linting

alextp commented on tensorflow/python/ops/image_grad.py:
```diff
@@ -22,6 +22,8 @@
 from tensorflow.python.framework import ops
 from tensorflow.python.ops import array_ops
 from tensorflow.python.ops import gen_image_ops
+from tensorflow.python.ops.math_ops import cast, add, reciprocal, square
```
In tf style we don't import individual symbols, we import the whole package and use the fully qualified name, like math_ops.cast, etc.

alextp commented on tensorflow/python/ops/image_grad.py:
```diff
@@ -153,3 +155,207 @@ def _CropAndResizeGrad(op, grad):
+  Returns:
+    x_reciprocal -> reciprocal of x added with my_eps
+  """
+  return reciprocal(x + my_eps)
```
This is not an ideal implementation, as it can turn -eps into 0, for example.

Instead, use math_ops.div_no_nan or tf.where to make this work correctly.

alextp reviewed (changes requested):
Can you add a test that the gradient is correct, using the gradient checker v2 from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/gradient_checker_v2.py ?

kmh4321 committed:
Review comments.

alextp commented:
Looks better! Just missing the gradient checker tests I mentioned earlier.

kmh4321 committed:
Add tests for rgb_to_hsv grad and fixes.

kmh4321 commented:
Hey @alextp 
I added a couple of tests however the gradient test doesn't pass. I am not sure if we can use the Jacobian method to verify gradients for such a complex function? 
I will re-check my derivation but I wanted to double check if the numerical method used to verify (ie Jacobian) can handle piecewise continuous functions with many conditions. 
https://en.wikipedia.org/wiki/HSL_and_HSV#From_RGB
This is the formula on which I calculated all the partial derivatives and as you can see there are many many sub-cases (MAX(r,g,b) and MIN(r,g,b) produce 6 ordered pairs for (max,min) and the derivative changes in each case) which need to be handled separately.

alextp commented:
The method we use to verify the gradients is just finite differences, so if we choose an area somewhat separate from the discontinuities it should be fine.

Another option to validate the gradient is to implement rgb_to_hsv in terms of basic TF ops (which is doable) and compare the result of tf.gradients there with the gradient implementation you provide.

kmh4321 commented:
Okay, thank you! For the second option did you mean implement rgb_to_hsv() in terms of basic TF ops or the gradient of rgb_to_hsv in basic tf ops? 
I'll try out both the suggestions and get back to you.

alextp commented:
I meant implementing rgb_to_hsv in terms of basic ops so we can get its
gradient also in terms of basic ops.

On Fri, Jun 28, 2019 at 1:10 PM Karthik Muthuraman <notifications@github.com>
wrote:

> Okay, thank you! For the second option did you mean implement rgb_to_hsv()
> in terms of basic TF ops or the gradient of rgb_to_hsv in basic tf ops?
> I'll try out both the suggestions and get back to you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/30152?email_source=notifications&email_token=AAABHRKHLXFDBXV6UZWY3U3P4ZV3FA5CNFSM4H3MUANKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY3BG6Q#issuecomment-506860410>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABHRNWYOKH5LZZX5QUJWTP4ZV3FANCNFSM4H3MUANA>
> .
>


-- 
 - Alex

gbaned commented:
@kmh4321 Did you get a chance to look on reviewer comments? Please let us know on the update. Thanks!

kmh4321 commented:
@gbaned Sorry I was out on vacation last week. Will take a look at this now.

frreiss commented on tensorflow/python/ops/image_grad.py:
```diff
@@ -153,3 +153,221 @@ def _CropAndResizeGrad(op, grad):
   return [grad0, grad1, None, None]
+
+
+def _CustomReciprocal(x):
```
After the most recent batch of changes, the docstring for this function no longer matches the arguments. It's probably best to drop this function definition and replace the calls to `_CustomReciprocal()` below with direct calls to `math_ops.div_no_nan`

kmh4321 replied:
Resolved in 032bdc6.

frreiss commented on tensorflow/python/ops/image_grad_test.py:
```diff
@@ -456,6 +458,38 @@ def testGradRandomBoxes(self):
+    def f(x):
+      return gen_image_ops.rgb_to_hsv(x)
+
+    x = np.random.rand(2, 20, 30, 3).astype(np.float32)
```
Initializing `x` with random values is likely to cause problems. If any of the (2*20*30) RGB values in the tensor happen to fall near a discontinuity, the test will fail. Better to use a constant value for `x` that still exercises the gradient function. There are a few ways to do this. You could manually pick a set of RGB values that exercise all the parts of your gradient equation; and include those values as a constant. Or you could add some code to detect and replace problematic RGB values.

kmh4321 replied:
Resolved in 032bdc6.

kmh4321 committed:
Formula fix, added tests.

kmh4321 commented:
@alextp  I have found the issue. I implemented it as per the definition I found on the wikipedia page, however the TF implementation has one final step where the "hues" are converted from degrees to [0, 1]. I have fixed that [here](https://github.com/tensorflow/tensorflow/pull/30152/commits/07fc1591f2e7e0394cf4a3f132043a05065b9595#diff-614af081276f42dbb98e5d731d7a1a86R283). 

As discussed above, I've also added a few test cases to avoid running into discontinuities and also added a [test](https://github.com/tensorflow/tensorflow/pull/30152/files#diff-bb1fdb8f5be7a85b4505589762579bf7R515) where I implement `rgb_to_hsv()` with native TF functions in a controlled input case (R>G>B). 

Now these tests are passing. Hope this is sufficient, will be glad to make any changes as required!

kmh4321 committed:
Fixed docstring for helper function.

alextp reviewed (approved):
Awesome! Thanks a lot for the contribution!
